### PR TITLE
Implement lakectl local clone

### LINKED ISSUE #6303: Implement lakectl local clone
Parent: https://github.com/treeverse/lakeFS/issues/6239

`lakectl local clone <lakefs path uri> [directory]`

Init directory with lakefs path and copy files to local

1. Dereference <lakefs path uri> (if it’s a branch, get the current commit it’s pointing to)
2. Copy files in dereferenced commit + path to the target directory (current dir if not set)
3. Create a .lakefs_ref.yaml file at the root of the local clone directory, with at_head set to the dereferenced commit ID
4. Find a .gitignore file in current directory or nearest parent and add:
  a. path/*
  b. !path/.lakefs_ref.yaml
7. This will ensure that this directory is not tracked in Git, except for its ref. Anyone cloning this repository will be able to lakectl local checkout that directory and get the required data

Optional flags:

- --presign - defaults to true. Will use pre-signed URLs to copy data without going through lakeFS server
- --parallelism - defaults to 25. Will run this many parallel downloads

Implementation notes:

- When copying the files, it is important to set the correct mtime on it. The mtime to use should be:
  - If the object has a x-client-mtime metadata key, it should be a unix timestamp. Use it
  - Otherwise, use the native Mtime returned for the object
- The reason for using x-client-mtime: The PutObject API has no means of setting a custom mtime on the object, but setting tags is possible.

### PR DESCRIPTION
**Pull request by N-o-Z**
### Linked Issue

Closes #6303

---

### Background

Implement lakectl local command which inits and downloads contents from lakefs path to a local directory
      
### New Feature

Part of task https://github.com/treeverse/lakeFS/issues/6239

### Testing Details

How were the changes tested?
Unit testing for tools, lakectl tests will be added for esti upon feature completion. SyncManager will be tested as part of the command (was tested manually)

### Breaking Change?

No

### CODE REVIEW

**Review by guy-har — CHANGES_REQUESTED** [review]
Nice! looks good
Added mostly suggestions, requesting changes due to the fact we can get stuck waiting on the error channel

File: pkg/uri/parser.go [hunk 1]
```diff
@@ -136,3 +136,11 @@ func Must(u *URI, e error) *URI {
...
 	return u
 }
+
+func WithRef(remote *URI, ref string) *URI {
+	return &URI{
+		Repository: remote.Repository,
+		Ref:        ref,
+		Path:       remote.Path,
+	}
+}
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/uri/parser.go:146]: Add documentation and made it a member function

File: cmd/lakectl/cmd/local_clone.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+const (
+	localCloneMinArgs = 1
+	localCloneMaxArgs = 2
+	filesChanSize     = 1000
+)
+
+var localCloneCmd = &cobra.Command{
+	Use:   "clone <path uri> [directory]",
+	Short: "init local directory and download objects from lakeFS path",
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply cmd/lakectl/cmd/local_clone.go:24]: How about "Clone a path from a lakeFS repository into a new directory."?

File: pkg/local/sync.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		ctx:            ctx,
+		client:         client,
+		httpClient:     http.DefaultClient,
+		progressBar:    NewProgressPool(),
+		maxParallelism: maxParallelism,
+		presign:        presign,
+	}
+}
+
+func (s *SyncManager) Sync(rootPath string, remote *uri.URI, changeSet <-chan *Change) error {
```
Comment (guy-har) [inline pkg/local/sync.go:90]: Can you please Document this method?

**Comment by guy-har** [comment]
Another thing I noticed is that the clone command fails when running on a branch with uncommitted changes

File: cmd/lakectl/cmd/local_clone.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	Args:  cobra.RangeArgs(localCloneMinArgs, localCloneMaxArgs),
+	Run: func(cmd *cobra.Command, args []string) {
+		remote := MustParsePathURI("path", args[0])
+		dir := "."
+		if len(args) == localCloneMaxArgs {
+			dir = args[1]
+		}
+		force := Must(cmd.Flags().GetBool("force"))
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply cmd/lakectl/cmd/local_clone.go:32]: Done

File: cmd/lakectl/cmd/local_clone.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+				listResp, err := client.ListObjectsWithResponse(cmd.Context(), remote.Repository, remote.Ref, &api.ListObjectsParams{
+					After:        (*api.PaginationAfter)(swag.String(after)),
+					Prefix:       (*api.PaginationPrefix)(remote.Path),
+					UserMetadata: swag.Bool(true),
+				})
+				if err != nil {
+					DieErr(err)
+				}
+				if listResp.HTTPResponse.StatusCode != http.StatusOK {
+					DieErr(fmt.Errorf("%w: HTTP %d", err, listResp.StatusCode()))
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply cmd/lakectl/cmd/local_clone.go:57]: Fixed

File: pkg/local/sync.go (hunk 3 above)
Comment (N-o-Z) [reply pkg/local/sync.go:90]: Done and refactored

File: pkg/local/sync.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	var wg sync.WaitGroup
+	for op := range changeSet {
+		<-ch // block until we have a slot
+		wg.Add(1)
+		go func(op *Change) {
+			err := s.apply(rootPath, remote, op)
+			if err != nil {
+				errCh <- err
+				return
+			}
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/local/sync.go:107]: Done

**Comment by guy-har** [comment]
@N-o-Z, another thing I'm concerned about is:
What do we expect to happen in case of failure (e.g. CTRL-C after a while)?
Currently, it's resulted in corrupted data without any way to identify it's a corrupted state, it just looks like we did some local changes. Maybe we should delete in case of failure, WDYT?

File: pkg/local/sync.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+		}
+
+		b := s.progressBar.AddReader(fmt.Sprintf("download %s", change.Path), sizeBytes)
+		barReader := b.Reader(body)
+		defer b.Done()
+
+		_, err = io.Copy(f, barReader)
+		if err != nil {
+			return fmt.Errorf("could not write file '%s': %w", destination, err)
+		}
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/local/sync.go:237]: Done

File: pkg/local/progress.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+func (p *ProgressSpinner) Done() {
+	p.t.Increment(progressTrackerDone)
+}
+
+type ProgressPool struct {
+	pw   progress.Writer
+	done chan bool
+}
+
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/local/progress.go:60]: progress.Writer is the progress pool AFAIK. We wrapped the trackers in order to provide additional functionality (e.g. ProgressUpdaterReader)

**Comment by N-o-Z** [comment]
> @N-o-Z, another thing I'm concerned about is: What do we expect to happen in case of failure (e.g. CTRL-C after a while)? Currently, it's resulted in corrupted data without any way to identify it's a corrupted state, it just looks like we did some local changes. Maybe we should delete in case of failure, WDYT?

Thanks, 
I'm not sure we want to complicate things by creating an "revert sync" flow.
I will open a separate issue to deal with interrupts for all local commands and provide a meaningful error message, similarly to what we did in import.

File: pkg/local/sync.go [hunk 9]
```diff
@@ -87,51 +87,50 @@ func NewSyncManager(ctx context.Context, client *api.ClientWithResponses, maxPar
 	}
 }
 
+// Sync - sync changes between remote and local directory given the Changes channel.
+// For each change, will apply download, upload or delete according to the change type and change source
 func (s *SyncManager) Sync(rootPath string, remote *uri.URI, changeSet <-chan *Change) error {
```
Comment (guy-har) [inline pkg/local/sync.go]: I have a confession:
I asked chatGPT if he can improve the documentation. Expected it to make slight changes, got a bit more than 
that, but I had to put it out here:

```
// Sync synchronizes changes between a local directory and a remote URI.
// The function consumes changes provided through the changeSet channel.
// Depending on the type and source of each change, the Sync function will
// either download, upload, or delete the associated content.
// 
// Parameters:
// - rootPath: The local directory path that should be kept in sync with the remote.
// - remote: The URI representing the remote directory.
// - changeSet: A channel through which changes to be synchronized are sent.
//
// The function leverages goroutines for parallel processing of changes, and the 
// maximum number of parallel operations is determined by s.maxParallelism. The synchronization 
// process is stopped early if an error is encountered.
// 
// After syncing, any empty directories within rootPath are pruned.
// 
// Returns:
// - nil if the sync completes successfully.
// - An error if any operation during the sync or pruning process fails.
```
It isn't aligned with our documentation, so I'm not sure what I'd suggest, but still needed to share 🤷

**Review by guy-har — APPROVED** [review]

**Pull request merged**